File: Programmers/C++/순위 검색_3.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#define FOR(i,n) for(int i=0;i<(n);++i)
#define FORS(i,s,n) for(int i=(s);i<(n);++i)

using namespace std;

vector<int> res;
unordered_map<string, vector<int>> um;

vector<int> solution(vector<string> info, vector<string> query) {
    
    for(auto s: info) {
        vector<string> v;
        string cur = "";
        
        for(auto c: s) {
            if(c==' ') {
                v.push_back(cur);
                cur="";
            }
            else cur+=c;
        }
        v.push_back(cur);
        
        int score = stoi(v[4]);
        
        um[v[0]+v[1]+v[2]+v[3]].push_back(score);
        
        um["-"+v[1]+v[2]+v[3]].push_back(score);
        um[v[0]+"-"+v[2]+v[3]].push_back(score);
        um[v[0]+v[1]+"-"+v[3]].push_back(score);
        um[v[0]+v[1]+v[2]+"-"].push_back(score);
        
        um["--"+v[2]+v[3]].push_back(score);
        um["-"+v[1]+"-"+v[3]].push_back(score);
        um["-"+v[1]+v[2]+"-"].push_back(score);
        um[v[0]+"--"+v[3]].push_back(score);
        um[v[0]+"-"+v[2]+"-"].push_back(score);
        um[v[0]+v[1]+"--"].push_back(score);
        
        um["---"+v[3]].push_back(score);
        um["--"+v[2]+"-"].push_back(score);
        um["-"+v[1]+"--"].push_back(score);
        um[v[0]+"---"].push_back(score);
        
        um["----"].push_back(score);
    }
    
    for(auto s: query) {
        vector<string> v;
        string cur = "";
        
        for(auto c: s) {
            if(c==' ') {
                if(cur!="and") v.push_back(cur);
                cur="";
            }
            else cur+=c;
        }
        v.push_back(cur);
        
        string str = v[0]+v[1]+v[2]+v[3];
        int sum = 0;
        int score = stoi(v[4]);
        for(auto x: um[str]) {
            if(x>=score) ++sum;
        }
        res.push_back(sum);
    }
    
    return res;
}

File: Programmers/Swift/순위 검색_3.swift
import Foundation

var hash = [String : [Int]]()
var sorted = [String : Bool]()
var res = [Int]()

func dfs(_ idx:Int, _ cnt:Int, _ v:[String], _ ary:inout [Int]) {
    if ary.count == cnt {
        var chk = [Bool](repeating:false, count:4)
        for i in ary {chk[i]=true}
        var str = ""
        let score = Int(v[4])!
        for i in 0..<4 {
            if chk[i] == true {str += v[i]}
            else {str += "-"}
        }
        str = String(str)
        if hash[str] == nil {
            hash[str] = [Int]()
        }
        hash[str]!.append(score)
        return
    }
    if(idx==4) {return}
    ary.append(idx)
    dfs(idx+1, cnt, v, &ary)
    ary.removeLast()
    dfs(idx+1, cnt, v, &ary)
}

func solution(_ info:[String], _ query:[String]) -> [Int] {
    
    for str in info {
        var v = str.split(separator:" ").map{String($0)}
        let score = Int(v[4])!
        var ary = [Int]()
        for cnt in 0...4 {
            dfs(0,cnt,v,&ary)
        }
    }
    
    for str in query {
        var v = str.split(separator:" ").filter{$0 != "and"}
        
        let str = String(v[0]+v[1]+v[2]+v[3])
        if hash[str] == nil {
            res.append(0)
            continue
        }
        if sorted[str] == nil {
            hash[str]!.sort()
            sorted[str] = true
        }
        let num = Int(v[4])!
        
        //binary search
        var mx = -1
        var l = 0
        var r = hash[str]!.count-1
        while l <= r {
            let mid = (l+r)/2
            if hash[str]![mid] < num {
                mx = max(mx, mid)
                l = mid+1
            } else {
                r = mid-1
            }
        }
        res.append(hash[str]!.count-mx-1)
    }
    return res
}
